keps/sig-storage/20200427-csi-driver-skip-permission.md

```diff
@@ -0,0 +1,143 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before bind mounting a CSI volume for a Pod, we use imprecise heuristics,
+such as presence of fsType on the PVC to determine if the volume supports fsGroup based
+permission change. These heuristics are known to be fragile, and cause problems with different
+storage types.
+
+To solve this issue we will add a new field called `csidriver.spec.SupportsFSGroup` 
+that allows the driver to define if it supports volume ownership modifications via
+fsGroup.
+
+## Motivation
+
+When a volume is mounted on the node, we modify volume ownership 
+before bind mounting the volume inside container; however, if the volume
+does not support these operations (such as NFS), then the change is still attempted. 
+This results in errors being reported to the user if the volume doesn't 
+support these operations.
+
+### Goals
+
+ - Allow CSI Drivers to opt-in to volume ownership changes.
+
+### Non-Goals
+
+ - This does not attempt to update existing CSI Drivers themselves.
+
+## Proposal
+
+We propose that CSI Drivers include a field that defines if the volume provided by the driver 
+supports changing volume ownership.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Currently Volume permission and ownership change is done automatically for every volume. 
+As part of this proposal we will change the algorithm that modifies volume ownership and permissions
+for CSI Drivers to check the new field, and skip volume ownership modifications if it is found to be
+`false`.
+
+When defining a CSI Driver, we propose that `csidriver.spec` be expanded to include a new field entitled 
+`SupportsFSGroup` which can have following possible values:
+
+ - true --> Current behavior. Attempt to modify the volume ownership and permissions to the defined `fsGroup` when the volume is attached.
+ - false --> New behavior. Attach the volume without attempting to modify volume ownership or permissions.
+
+
+```go
+type CSIDriverSpec struct {
+    // SupportsFSGroup ← new field
+    // Defines if the underlying volume supports changing ownership and 
+    // permission of the volume before being attached. 
+    // If set to true, SupportsFSGroup indicates that 
+    // the underlying volume supports volume ownership and permission
+    // changes, and the filesystem will be modified by the defined fsGroup.
+    // If set to false, then the volume will be attached without modifying
+    // the volume's ownership or permissions.
+    // Defaults to `true`.
+    // + optional
+    SupportsFSGroup *bool
```

Comment:
I think - we can't simply default this to true because current behaviour is more nuanced - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L379 .

I think this field should ideally be an enum with at least three values. First one:
- detectFSGroupSupport (may be - I am not very sure about the name) - which is current behaviour without this enhancement.
- fsGroupUnsupported (volume type does not supports fsgroup)
- fsGroupSupported (volume type does support fsGroup)

Reply:
I am not very sure about these names yet. We need to think more.

Reply:
What if we change this slightly so that the field is `csidriver.spec.ApplyFSGroup`, and the values are:

- Always - equal to your `fsGroupSupported`
- Never - equal to your `fsGroupUnsupported`
- OnDemand - equal to `detectFSGroupSupport`

This would have similar behavior to the above, but would be a bit clearer. It would also align with some of the other names we have for similar enums.

Reply:
cc @msau42

Reply:
discussed this offline. Using boolean pointer works as well, but again it should default to `nil` value rather than `true`.

Reply:
This has been updated to reference the three modes with a `bool` - nil/true/false.

Reply:
Actually I like the idea of making the "nil" behavior more explicit with a type. For someone coming in without any background on this behavior, I think it would be surprising that the nil behavior did what it currently does. And if we ever have a v2, then it's very obvious what behaviors may need to be maintained.

Reply:
I'm marking this conversation as unresolved. I've updated the KEP to have an explicit type for controlling this behavior instead of using a boolean pointer. Let me know your thoughts on these names, and this approach.